Do not record tool call id if no tool has been called

sysradium opened the pull request:
I came across this issue here: https://github.com/huggingface/smolagents/issues/570

The problem is that if you have callbacks which modify observations like it is done here https://github.com/huggingface/smolagents/blob/main/src/smolagents/vision_web_browser.py#L66

It may happen that there is no tool call recorded yet and hence `self.tool_calls` is `None`. (ref https://github.com/sysradium/smolagents/blob/web-tool-call-issue/src/smolagents/agents.py#L871)

For example in my case it happened with `vision_web_browser.py` when the model did not return any Python code on the first call to it. That led to the exception here https://github.com/sysradium/smolagents/blob/web-tool-call-issue/src/smolagents/agents.py#L867 which happens right before we set `memory.step.tools_calls`.

So essentially you end up having `data:,` as an observation but no recorded tool call.

Maybe not the best solution, but I decided to just not print `Call id: ...` if there has been no successful tool call.

touseefahmed96 commented on src/smolagents/memory.py:
```diff
@@ -106,7 +106,7 @@ def to_messages(self, summary_mode: bool = False, show_model_input_messages: boo
                         {
                             "type": "text",
-                            "text": f"Call id: {self.tool_calls[0].id}\nObservation:\n{self.observations}",
+                            "text": f"f{'Call id: {self.tool_calls[0].id}\n' if self.tool_calls else ''}Observation:\n{self.observations}",
```
why this extra f?

sysradium replied:
Because there is an if-statement inside the outer f-string.

sysradium commented:
@aymeric-roucher another one here if you have some time.

aymeric-roucher commented on src/smolagents/memory.py:
```diff
@@ -106,7 +106,7 @@ def to_messages(self, summary_mode: bool = False, show_model_input_messages: boo
                         {
                             "type": "text",
-                            "text": f"Call id: {self.tool_calls[0].id}\nObservation:\n{self.observations}",
+                            "text": f"f{'Call id: {self.tool_calls[0].id}\n' if self.tool_calls else ''}Observation:\n{self.observations}",
```
I think this would be more readable:
```suggestion
                            "text": f"Call id: {self.tool_calls[0].id}\n" if self.tool_calls else "" + f"Observation:\n{self.observations}",
```

sysradium commented:
@aymeric-roucher sure, no will try to figure out a way.

sysradium commented:
@aymeric-roucher 

Added a test fixing some things along the way:

1. Added `mlx` into dev deps. Otherwise after running `pip install -e ".[dev]" ` on a clean setup would get:
```py
FAILED tests/test_models.py::ModelTests::test_get_mlx_message_no_tool - ModuleNotFoundError: Please install 'mlx-lm' extra to use 'MLXModel': `pip install 'smolagents[mlx-lm]'`
```

2. Modified `Makefile` target so that you could overrde `python` command. For example if you use `uv` now you can do:
```makefile
PYTHON='uv run python' make test 
PYTHON='uv run python' make quality 
```
Or override it with python path you need. Just a bit of flexibility.

3. With a test fixture forced `MultiAgentsTests` to dump data into a tmpdir, otherwise you are getting unwanted guests:
```
>>> gst
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        agent_export/
```

4. Some minor type hint fixes.

sysradium commented:
@g-eoj I don't think so. You can check here https://pypi.org/project/mlx/#files that they have prebuilt binaries for linux and mac. As well as I found that it has a windows support https://github.com/ml-explore/mlx/issues/1513

@aymeric-roucher a kind ping

sysradium commented:
@albertvillanova rebased the  change.

sysradium commented:
@aymeric-roucher rebased

albertvillanova reviewed (commented):
Thanks.

The rebase seems did not work properly. I'm cherry-picking the relevant commits.

sysradium commented:
@albertvillanova what issues are there? I can't see conflicts atm
